Update docs to include how to run a pack with path (#9570)
github/codeql: docs/codeql/codeql-cli/analyzing-databases-with-the-codeql-cli.rst

The ``analyze`` command above runs the default suite from ``microsoft/coding-standards v1.0.0`` and the latest version of ``github/security-queries`` on the specified database.
For further information about default suites, see ":ref:`Publishing and using CodeQL packs <publishing-and-using-codeql-packs>`".

Running a subset of queries in a CodeQL pack
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> Additionally, you can include a path at the end of a pack specification to run a subset of queries inside the pack. This applies to any command that locates or runs queries within a pack.

Reviewer: @aeisenberg, i actually missed that we didn't call out which version of the CLI supports this functionality. Can we add it in this section please? (It'll look a bit messy, but it's important for users who are tied to older CLI versions for some reason.)

The complete way to specify a set of queries is in the form ``scope/name@range:path``, where:

- ``scope/name`` is the qualified name of a CodeQL pack.
- ``range`` is a `semver range <https://docs.npmjs.com/cli/v6/using-npm/semver#ranges>`_.
- ``path`` is a file system path to a single query, a directory containing queries, or a query suite file.

> If a ``scope/name`` is specified, the ``range`` and ``path`` are
> optional. A missing ``range`` implies the latest version of the
> specified pack. A missing ``path`` implies the default query suite
> of the specified pack.

Reviewer 2: Suggest:

```suggestion
When you specify a ``scope/name``, the ``range`` and ``path`` are
optional. If you omit a ``range`` then the latest version of the
specified pack is used. If you omit a ``path`` then the default query suite
of the specified pack is used.
```

The ``path`` can be one of a ``*.ql`` query file, a directory

> containing one or more queries, or a ``.qls`` query suite file. If
> there is no pack name specified, then a ``path`` must be provided,
> and will be interpreted relative to the current working directory
> of the current process.

Reviewer 2: ```suggestion
containing one or more queries, or a ``.qls`` query suite file. If
you omit a pack name, then you must provide a ``path``,
which will be interpreted relative to the working directory
of the current process.
```

> If a ``scope/name`` and ``path`` are specified, then the ``path`` cannot
> be absolute. It is considered relative to the root of the CodeQL
> pack.

Reviewer 2: Suggest:

```suggestion
If you specify a ``scope/name`` and ``path``, then the ``path`` cannot
be absolute. It is considered relative to the root of the CodeQL
pack.

```

- ``scope/name`` is the qualified name of a CodeQL pack.
- ``range`` is a `semver range <https://docs.npmjs.com/cli/v6/using-npm/semver#ranges>`_.
- ``path`` is a file system path to a single query, a directory containing queries, or a query suite file.

If a ``scope/name`` is specified, the ``range`` and ``path`` are
optional. A missing ``range`` implies the latest version of the
specified pack. A missing ``path`` implies the default query suite
of the specified pack.

The ``path`` can be one of a ``*.ql`` query file, a directory
containing one or more queries, or a ``.qls`` query suite file. If
there is no pack name specified, then a ``path`` must be provided,
and will be interpreted relative to the current working directory
of the current process.

If a ``scope/name`` and ``path`` are specified, then the ``path`` cannot
be absolute. It is considered relative to the root of the CodeQL
pack.

The relevant commands are:

* `codeql database analyze <../manual/database-analyze>`__.
* `codeql database run-queries <../manual/database-run-queries>`__.
* `codeql execute queries <../manual/execute-queries>`__.

> * `codeql resolve queries <../manual/resolve-queries>`__.

Reviewer: These are plumbing commands that we generally don't expose in these procedural docs. If anyone is interested in they can poke around in the manual.

> 
> The relevant commands are:
> 
> * `codeql database analyze <../manual/database-analyze>`__.
> * `codeql database run-queries <../manual/database-run-queries>`__.
> * `codeql execute queries <../manual/execute-queries>`__.
> * `codeql resolve queries <../manual/resolve-queries>`__.

Reviewer: ```suggestion

```

> 
> For example::
> 

Reviewer: ```suggestion

```

Reviewer: The `::` here is causing the doc check to fail as it's expecting line 166 to be indented. If you commit this suggestion the test _should_ go green.

>     # Analyze a database using all queries in the experimental/Security folder within the codeql/cpp-queries
>     # CodeQL query pack.
>     codeql database analyze --format=sarif-latest --output=results <db> \
>         codeql/cpp-queries:experimental/Security
> 
>     # Analyse using only the RedundantNullCheckParam.ql query in the codeql/cpp-queries CodeQL query pack.
>     codeql database analyze --format=sarif-latest --output=results <db> \
>         'codeql/cpp-queries:experimental/Likely Bugs/RedundantNullCheckParam.ql'
> 
>     # Analyse using the cpp-security-and-quality.qls query suite in the codeql/cpp-queries CodeQL query pack.
>     codeql database analyze --format=sarif-latest --output=results <db> \
>         'codeql/cpp-queries:codeql-suites/cpp-security-and-quality.qls'
> 
>     # Analyse using the cpp-security-and-quality.qls query suite from a version of the codeql/cpp-queries pack
>     # that is >= 0.0.3 and < 0.1.0 (the highest compatible version will be chosen).
>     # All valid semver ranges are allowed. See https://docs.npmjs.com/cli/v6/using-npm/semver#ranges
>     codeql database analyze --format=sarif-latest --output=results <db> \
>        'codeql/cpp-queries@~0.0.3:codeql-suites/cpp-security-and-quality.qls'
> 

Reviewer: Tidied up the examples here a bit. 

```suggestion
To analyze a database using all queries in the `experimental/Security` folder within the `codeql/cpp-queries` CodeQL pack you can use::

    codeql database analyze --format=sarif-latest --output=results <db> \
        codeql/cpp-queries:experimental/Security

To run the `RedundantNullCheckParam.ql` query in the `codeql/cpp-queries` CodeQL pack use::

    codeql database analyze --format=sarif-latest --output=results <db> \
        'codeql/cpp-queries:experimental/Likely Bugs/RedundantNullCheckParam.ql'

To analyze your database using the `cpp-security-and-quality.qls` query suite from a version of the `codeql/cpp-queries` CodeQL pack that is >= 0.0.3 and < 0.1.0 (the highest compatible version will be chosen) you can use::

    codeql database analyze --format=sarif-latest --output=results <db> \
       'codeql/cpp-queries@~0.0.3:codeql-suites/cpp-security-and-quality.qls'
```

Running a subset of queries in a CodeQL pack
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> Additionally, CodeQL CLI v2.10.0 or later, you can include a path at the end of a pack specification to run a subset of queries inside the pack. This applies to any command that locates or runs queries within a pack.

Reviewer: Is this correct? The [changelog entry](https://github.com/github/codeql-cli-binaries/blob/main/CHANGELOG.md#release-281-2022-02-15) suggests that this had been available since 2.8.1.

Author: You're right. Let me fix.

```suggestion
Additionally, CodeQL CLI v2.8.1 or later, you can include a path at the end of a pack specification to run a subset of queries inside the pack. This applies to any command that locates or runs queries within a pack.
```
